### DISCUSSION TITLE
LaTeX support

### DISCUSSION THREAD

**Discussion by rjacquinot** [gonzalezreal/swift-markdown-ui#208] [Ideas]
Hi, 

LaTeX doesn't appear to be supported yet. It would be great to see it added. 

Do you have any future plans to support it?

Thanks

**Comment by gonzalezreal** [comment]
Hi @rjacquinot,

I'm sorry, but I don't have any plans to support LaTeX right now. However, I welcome suggestions for integrating a third-party library that provides LaTeX rendering in SwiftUI.

**Reply by rjacquinot** [reply]
Thanks for your reply. If I come across any great libraries that could potentially meet this need, I will be sure to share them with you.

**Comment by samrat** [comment]
Has there been any further developments on this front? I'm interested in integrating the functionalities of https://github.com/colinc86/LaTeXSwiftUI and this library. 

Any advice on how I might go about doing that?

**Reply by gonzalezreal** [reply]
Hi @samrat,

I still haven't got the time to look into this. We would need a library that renders LaTeX into a SwiftUI `Text` view to support inline mathematical expressions inside a text paragraph. Supporting block mathematical expressions should be more accessible.

This is an interesting feature, and I will look into it as soon as possible.

**Comment by Abel-Ming** [comment]
Hi,

Is there any progress on LaTex support right now, I'm interested in this, it would be great if LaTex presentations were supported in this library.

Thanks

**Reply by gonzalezreal** [reply]
Hi,
I am currently busy with other stuff and have no plans for LaTeX support in the short term.

**Reply by Abel-Ming** [reply]
Thank you very much for your reply.

**Comment by Asilbek25** [comment]
Hi @gonzalezreal,

Could I possibly offer a payout to add LaTeX support to this project? I've tried other alternatives, but they are not as smooth and functional as Markdown UI, and they lack support for other important features. 

I would greatly appreciate an answer. If you have any questions, please feel free to contact me at abduwdeveloper@gmail.com.

Thank you!

**Reply by gonzalezreal** [reply]
I am really sorry but currently I have no time to embark in such a project.

**Reply by Asilbek25** [reply]
I understand, thanks for the reply

**Reply by Visual-Studio-Coder** [reply]
hopefully the open-source community can pull through 🤞🏼

**Comment by rafaelpedretti-toast** [comment]
I was able to use it as images from https://latex.codecogs.com/
In my use case I get latex responses from a LLM so I ask it to wrap any latex expression in `<latex>` tag and in the app I apply the transformation to use the expression as a markdown image, something like `[](https://latex.codecogs.com/png.latex?\\dpi{200}<transformed expression>)`

I use png because iOS does not support SVG out of the box and SVG works just fine. You can adjust the DPI to your needs.
The code below are the image providers to enable dark/light mode and scaling to dynamic type
```swift
struct ThemeAwareImageProvider: ImageProvider {
    let scaleFactor: CGFloat

    public func makeImage(url: URL?) -> some View {
        NetworkImage(url: url) { state in
            switch state {
            case .failure, .empty:
                Color.clear
                    .frame(width: 0, height: 0)
            case let .success(image, idealSize):
                image
                    .renderingMode(.template)
                    .resizable()
                    .aspectRatio(contentMode: .fit)
                    .padding(.leading, .x1)
                    .frame(
                        idealWidth: idealSize.width * scaleFactor / 2,
                        idealHeight: idealSize.height * scaleFactor / 2
                    )
            }
        }
    }
}

struct ThemeAwareInlineImageProvider: InlineImageProvider {
    let scaleFactor: CGFloat

    public func image(with url: URL, label: String) async throws -> Image {
        let image = try await Image(
            DefaultNetworkImageLoader.shared.image(from: url),
            scale: 2 / scaleFactor,
            label: Text(label)
        )

        return image.renderingMode(.template).resizable()
    }
}
```

Here's the code to transform the `<latex>expression</latex>` into a markdown image:
```swift
extension String {
    func withLatexImage(dpi: Int = 200) -> String {
        let regexp = try? Regex("<latex>\\s*(.+?)\\s*</latex>")
        guard let regexp else { return self }
        let transformed = self.replacing(regexp, with: { match in
            guard let latexContent = match.output[1].substring else {
                return ""
            }

            let queryParameter = latexContent.replacing(" ", with: "&space;")

            return "![](https://latex.codecogs.com/png.latex?\\dpi{\(dpi)}\(queryParameter))"
        })

        return transformed
    }
}

```

![image](https://github.com/user-attachments/assets/62a0d663-9868-4045-90fa-f8a3298a2a8b)

**Comment by gonzalezreal** [comment]
Hi everyone — quick update on this discussion.

LaTeX / math expression rendering is now supported in [Textual](https://github.com/gonzalezreal/textual), my newer SwiftUI-native text rendering engine, including inline and display math.

Context: https://github.com/gonzalezreal/swift-markdown-ui/discussions/437

Hope this helps, and thanks for the patience!

**Reply by rjacquinot** [reply]
Amazing! Thanks so much for your hard work.
